lib/proxy_api/external_ipam.rb

```diff
@@ -0,0 +1,127 @@
+    # Returns: Hash with next_ip on success, hash with error otherwise
+    #   Example(success): {"cidr":"100.10.10.0/24","next_ip":"100.10.10.23"}
+    #   Example(error):   {"error":"The specified subnet does not exist in phpIPAM."}
+    def next_ip(subnet)
+      cidr = subnet.network + '/' + subnet.cidr.to_s if subnet.network.present?
+      get("/next_ip?cidr=#{cidr}")
+    rescue => e
+      raise ProxyException.new(url, e, N_("Unable to retrieve the next available IP for subnet %s from phpIPAM."), cidr)
+    end
+
+    # Adds the specified IP address to the specified subnet in phpIPAM
+    # Returns: Hash with message on success, hash with error otherwise
+    #   Example(success): {"message":"IP 100.10.10.123 added to subnet 100.10.10.0/24 successfully."}
+    #   Example(error):   {"error":"The specified subnet does not exist in phpIPAM."}
+    def add_ip_to_subnet(ip, subnet)
+      cidr = subnet.network + '/' + subnet.cidr.to_s if subnet.network.present?
+      parse(post({:ip => ip, :cidr => cidr}, "/add_ip_to_subnet"))
+    rescue => e
+      raise ProxyException.new(url, e, N_("Unable to add IP %s to the subnet %s in phpIPAM."), ip, cidr)
+    end
+
+    # Get a list of phpIPAM sections
+    # Returns: Array of sections(as json) on success, hash with error otherwise
+    #   Example(success): [{"id":"5","name":"Awesome Section","description":"A totally awesome Section",
+    #                       "masterSection":"0","permissions":"[]","strictMode":"1","subnetOrdering":"default",
+    #                       "order":null,"editDate":"2019-04-19 21:49:55","showVLAN":"1","showVRF":"1",
+    #                       "showSupernetOnly":"1","DNS":null}]]
+    #   Example(error):   {"error":"Unable to connect to phpIPAM server"}
+    def get_sections
+      get("/sections")
+    rescue => e
+      raise ProxyException.new(url, e, N_("Unable to obtain sections from phpIPAM."))
+    end
+
+    # Get a list of subnets for given phpIPAM section
+    # Returns: Array of subnets(as json) on success, hash with error otherwise
+    #   Example(success): Some less common fields removed for brevity
+    # {
+    #   "code":200,
+    #   "success":true,
+    #   "data":[
+    #     {
+    #       "id":"9",
+    #       "subnet":"100.10.10.0",
+    #       "mask":"24",
+    #       "sectionId":"5",
+    #       "description":"Full Subnet",
+    #       ...
+    #     },
+    #     {
+    #       "id":"10",
+    #       "subnet":"100.20.20.0",
+    #       "mask":"24",
+    #       "sectionId":"5",
+    #       "description":"Open Subnet",
+    #       ...
+    #     }
+    #   ],
+    #   "time":0.021
+    # }
+    #   Example(error):   {"error":"Unable to connect to phpIPAM server"}
+    def get_subnets(section_id)
+      get("/sections/#{section_id}/subnets")
+    rescue => e
+      raise ProxyException.new(url, e, N_("Unable to obtain subnets for section_id %s from phpIPAM."), section_id.to_s)
+    end
```

Comment:
Consider using just section name instead of id if its unique. Also return subnets only in simplest format, no extra fields only those which are really necessary. Make this a generic API.

Reply:
We could use section name(it is unique), however this would mean an extra API call to get the id of the section, which in turn would be used to get the subnets collection. Unfortunately the phpIPAM API doesn't support getting the subnets be section name.

Reply:
For the sake of clean generic external IPAM API I think it is probably worth it. However I don't insist.

What you should probably do is to design the Smart Proxy API in a way that sections are optional - this call should also work without any section provided. In that case, your implementation should probably raise an error "section is required" or something, but other implementation should go through.